src/gfx/convert.cpp

```diff
@@ -0,0 +1,1001 @@
+	// TODO: allow picking one of several solvers?
+	auto [mappings, nbPalettes] = packing::overloadAndRemove(protoPalettes);
+	assert(mappings.size() == protoPalettes.size());
+
+	if (options.verbosity >= Options::VERB_INTERM) {
+		fprintf(stderr, "Proto-palette mappings: (%zu palette%s)\n", nbPalettes,
+		        nbPalettes != 1 ? "s" : "");
+		for (size_t i = 0; i < mappings.size(); ++i) {
+			fprintf(stderr, "%zu -> %zu\n", i, mappings[i]);
+		}
+	}
+
+	std::vector<Palette> palettes(nbPalettes);
+	// Generate the actual palettes from the mappings
+	for (size_t protoPalID = 0; protoPalID < mappings.size(); ++protoPalID) {
+		auto &pal = palettes[mappings[protoPalID]];
+		for (uint16_t color : protoPalettes[protoPalID]) {
+			pal.addColor(color);
+		}
+	}
+
+	// "Sort" colors in the generated palettes, see the man page for the flowchart
+	auto [embPalSize, embPalRGB, embPalAlpha] = png.getEmbeddedPal();
+	if (embPalRGB != nullptr) {
+		sorting::indexed(palettes, embPalSize, embPalRGB, embPalAlpha);
+	} else if (png.isSuitableForGrayscale()) {
+		sorting::grayscale(palettes, png.getColors().raw());
+	} else {
+		sorting::rgb(palettes);
+	}
+	return {mappings, palettes};
+}
+
+static std::tuple<DefaultInitVec<size_t>, std::vector<Palette>>
+    makePalsAsSpecified(std::vector<ProtoPalette> const &protoPalettes, Png const &png) {
+	if (options.palSpecType == Options::EMBEDDED) {
+		// Generate a palette spec from the first few colors in the embedded palette
+		auto [embPalSize, embPalRGB, embPalAlpha] = png.getEmbeddedPal();
+		if (embPalRGB == nullptr) {
+			fatal("`-c embedded` was given, but the PNG does not have an embedded palette!");
+		}
+
+		// Fill in the palette spec
+		options.palSpec.emplace_back(); // A single palette, with `#00000000`s (transparent)
+		assert(options.palSpec.size() == 1);
+		// TODO: abort if ignored colors are being used; do it now for a friendlier error
+		// message
+		if (embPalSize > options.maxOpaqueColors()) { // Ignore extraneous colors if they are unused
+			embPalSize = options.maxOpaqueColors();
+		}
+		for (int i = 0; i < embPalSize; ++i) {
+			options.palSpec[0][i] = Rgba(embPalRGB[i].red, embPalRGB[i].green, embPalRGB[i].blue,
+			                             embPalAlpha ? embPalAlpha[i] : 0xFF);
+		}
+	}
+
+	// Convert the palette spec to actual palettes
+	std::vector<Palette> palettes(options.palSpec.size());
+	auto palIter = palettes.begin(); // TODO: `zip`
+	for (auto const &spec : options.palSpec) {
+		for (size_t i = 0; i < options.nbColorsPerPal; ++i) {
+			(*palIter)[i] = spec[i].cgbColor();
+		}
+		++palIter;
+	}
+
+	// Iterate through proto-palettes, and try mapping them to the specified palettes
+	DefaultInitVec<size_t> mappings(protoPalettes.size());
+	for (size_t i = 0; i < protoPalettes.size(); ++i) {
+		ProtoPalette const &protoPal = protoPalettes[i];
+		// Find the palette...
+		auto iter = std::find_if(palettes.begin(), palettes.end(), [&protoPal](Palette const &pal) {
```

Comment:
Just noting that this looks well-indented, which hasn't been the case for every lambda within a function call.